### PR TITLE
server: Add api v2 endpoints for dbs, tables, users, events

### PR DESCRIPTION
**Pull request by itsbilal**
This change adds api v2 endpoints for dbs, users, tables, and events.
These are backed by SQL queries, and support pagination as well
as v2 docs and auth.

Release note (general change): Add HTTP v2 endpoints for
querying databases, tables, users, and events in a db.

### CODE REVIEW

**Review by knz — COMMENTED** [review]
Code LGTM, but  I'd like to question the overall design of the "database details" API. 
Currently the size of a response payload can be arbitrary large: for example if there are many tables inside or many grants.

I think it would be good to explain in that API's documentation (and inline comments, towards the future maintainers of this) that this API is, in fact, misdesigned. And instead we will want:

- list databases
- list tables for a given database
- retrieve database grants for 1 database

etc

where each dataset that can have a dynamic size can be properly paginated.

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-:-MYoLd-r1nYddeNZYPuM:bgfibfv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @itsbilal)

___
*[docs/generated/swagger/spec.json, line 597 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MYoKqPu4DHXS8fc560-:-MYoKqPu4DHXS8fc5600:b-qjnp5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/013ae48e17a660483354e95710ebf372ff4a9512/docs/generated/swagger/spec.json#L597)):*
> ```JSON
>           "description": "Key is only set if this is a constraint on locality.",
>           "type": "string",
>           "x-go-name": "Key"
> ```

Is there a way to omit these `x-go` definitions from the spec?
___
*[pkg/server/api_v2_admin.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MYoLJUS99HwAJH-zEdA:-MYoLJUTBwco46lOjNCt:bke4uia) ([raw file](https://github.com/cockroachdb/cockroach/blob/013ae48e17a660483354e95710ebf372ff4a9512/pkg/server/api_v2_admin.go#L99)):*
> ```Go
> 		return
> 	}
> 	if limit > 0 && len(resp.Users) == limit {
> ```

I'd recommend `len(resp.Users) >= limit` in the condition and `Next = offset + len(resp.Users)` below

To cover the hypothetical case where the SELECT query above is changed to return more rows than requested.
___
*[pkg/server/api_v2_admin.go, line 177 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MYoLXYG1INgQOy6BQP-:-MYoLXYH3MurUlq0RJx_:b-z3p5ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/013ae48e17a660483354e95710ebf372ff4a9512/pkg/server/api_v2_admin.go#L177)):*
> ```Go
> 	}
> 	resp.EventsResponse = *eventsResp
> 	if limit > 0 && len(resp.Events) == limit {
> ```

ditto above

**Review by itsbilal — COMMENTED** [review]
TFTR! Added the comments on splitting up the database details API. Good point there.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-:-MYupafcC8PChICyGff1:b708cvk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @knz)

___
*[docs/generated/swagger/spec.json, line 597 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MYoKqPu4DHXS8fc560-:-MYupdzKAxeDzMjNniIu:b-x2mdyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/013ae48e17a660483354e95710ebf372ff4a9512/docs/generated/swagger/spec.json#L597)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Is there a way to omit these `x-go` definitions from the spec?
</blockquote></details>

I couldn't find a way to do that - seems to be a `go-swagger` internal thing.
___
*[pkg/server/api_v2_admin.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MYoLJUS99HwAJH-zEdA:-MYuq-5_AH-MnbIof8li:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/013ae48e17a660483354e95710ebf372ff4a9512/pkg/server/api_v2_admin.go#L99)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'd recommend `len(resp.Users) >= limit` in the condition and `Next = offset + len(resp.Users)` below

To cover the hypothetical case where the SELECT query above is changed to return more rows than requested.
</blockquote></details>

Done.
___
*[pkg/server/api_v2_admin.go, line 177 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MYoLXYG1INgQOy6BQP-:-MYuq-NR4_5oxLl7b_PA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/013ae48e17a660483354e95710ebf372ff4a9512/pkg/server/api_v2_admin.go#L177)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto above
</blockquote></details>

Done.

**Review by knz — COMMENTED** [review]
> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

**Review by itsbilal — COMMENTED** [review]
I've split the databaseDetails API into two others, one each for grants and tables, individually paginated. PTAL. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-:-MZ-WfZ4A9ffHoMLizuM:b-ii6cv6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @knz)

**Review by knz — COMMENTED** [review]
> Reviewed 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @itsbilal)

___
*[docs/generated/swagger/spec.json, line 85 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZIMIzU3xlIo5EDpOIJ:-MZIMIzU3xlIo5EDpOIK:b8su1px) ([raw file](https://github.com/cockroachdb/cockroach/blob/8abc2ebb0734126824e8d1c7b279c979b5a11f40/docs/generated/swagger/spec.json#L85)):*
> ```JSON
>     "/databases/{database}/grants/": {
>       "get": {
>         "description": "Grants are the results of SHOW GRANTS for this database.",
> ```

I would rephrase the description to avoid referring to `SHOW`. The description of the API should be self-contained so that:
- readers don't need to look in multiple places to understand how to use the API endpoint
- we should be able to guarantee that the API doesn't magically start reporting something different if SHOW evolves (i.e. our api guarantees are not dependent on changes elsewhere).

(if you've referred to SHOW in other api docs, the same applies to them)
___
*[docs/generated/swagger/spec.json, line 136 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZIMjCyCXmPfsEmymUA:-MZIMjCyCXmPfsEmymUB:b-w6ypv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8abc2ebb0734126824e8d1c7b279c979b5a11f40/docs/generated/swagger/spec.json#L136)):*
> ```JSON
>           {
>             "type": "string",
>             "description": "Name of database being looked up.",
> ```

Name of _the_ database being looked up.
___
*[pkg/server/api_v2_admin.go, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZINXc79Tsg0kIPn2aM:-MZINXc79Tsg0kIPn2aN:b-9vz3u3) ([raw file](https://github.com/cockroachdb/cockroach/blob/8abc2ebb0734126824e8d1c7b279c979b5a11f40/pkg/server/api_v2_admin.go#L1)):*
> ```Go
> // Copyright 2021 The Cockroach Authors.
> ```

`api_v2_sql_schema.go` instead of `api_v2_admin.go` perhaps?
___
*[pkg/server/api_v2_admin.go, line 301 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZINz6z6ngOBUgpcjFT:-MZINz6z6ngOBUgpcjFU:bv5w0bk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8abc2ebb0734126824e8d1c7b279c979b5a11f40/pkg/server/api_v2_admin.go#L301)):*
> ```Go
> 	resp := databaseDetailsResponse{
> 		DescriptorID:    dbDetailsResp.DescriptorID,
> 		ZoneConfig:      dbDetailsResp.ZoneConfig,
> ```

the `zonepb.ZoneConfig` type is not a safe type for a public API unfortunately, because of the `SubZoneSpans` sub-field.

I fear we can't just return the zone config for anything in these APIs using the `zonepb` protobufs. We'll want to do something else instead. My suggestion could be to report the zone config as a SQL pretty-printed `SET ZONE CONFIGURATION` statement.

You could evolve this PR in this direction, or dumb down this "get database details" API to just "get the database desc ID" and call it a day.
___
*[pkg/server/api_v2_admin_test.go, line 63 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZINinAD2ZS_Mn4-z4l:-MZINinBE7PxtTO2Sz1-:bxt8xq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/8abc2ebb0734126824e8d1c7b279c979b5a11f40/pkg/server/api_v2_admin_test.go#L63)):*
> ```Go
> 
> func TestDatabasesTablesV2(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> ```

This test suite feels like it should be converted to a `datadriven` test. The payloads would look good in expected test outputs.

I don't think you absolutely need to do this now (in this PR) but if you choose not to, please file an issue and outline for the benefit of the next person how this could look like.

**Review by itsbilal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @knz)

___
*[docs/generated/swagger/spec.json, line 85 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZIMIzU3xlIo5EDpOIJ:-MZOTUSX-I0K0KhL3tZf:byxrjyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8abc2ebb0734126824e8d1c7b279c979b5a11f40/docs/generated/swagger/spec.json#L85)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I would rephrase the description to avoid referring to `SHOW`. The description of the API should be self-contained so that:
- readers don't need to look in multiple places to understand how to use the API endpoint
- we should be able to guarantee that the API doesn't magically start reporting something different if SHOW evolves (i.e. our api guarantees are not dependent on changes elsewhere).

(if you've referred to SHOW in other api docs, the same applies to them)
</blockquote></details>

That makes sense to me! I was just copying over the definition from the protobuf. Done.
___
*[docs/generated/swagger/spec.json, line 136 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZIMjCyCXmPfsEmymUA:-MZOTZIz1ETTk-mz0d2I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8abc2ebb0734126824e8d1c7b279c979b5a11f40/docs/generated/swagger/spec.json#L136)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Name of _the_ database being looked up.
</blockquote></details>

Done.
___
*[pkg/server/api_v2_admin.go, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZINXc79Tsg0kIPn2aM:-MZOTZlC5y1IuJnwalTt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8abc2ebb0734126824e8d1c7b279c979b5a11f40/pkg/server/api_v2_admin.go#L1)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`api_v2_sql_schema.go` instead of `api_v2_admin.go` perhaps?
</blockquote></details>

Done.
___
*[pkg/server/api_v2_admin.go, line 301 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZINz6z6ngOBUgpcjFT:-MZOT_PJ8c2my_A0ZSmW:b-uk0ukr) ([raw file](https://github.com/cockroachdb/cockroach/blob/8abc2ebb0734126824e8d1c7b279c979b5a11f40/pkg/server/api_v2_admin.go#L301)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

the `zonepb.ZoneConfig` type is not a safe type for a public API unfortunately, because of the `SubZoneSpans` sub-field.

I fear we can't just return the zone config for anything in these APIs using the `zonepb` protobufs. We'll want to do something else instead. My suggestion could be to report the zone config as a SQL pretty-printed `SET ZONE CONFIGURATION` statement.

You could evolve this PR in this direction, or dumb down this "get database details" API to just "get the database desc ID" and call it a day.

</blockquote></details>

Good point, for now I just reduced scope to just descriptor ID.
___
*[pkg/server/api_v2_admin_test.go, line 63 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZINinAD2ZS_Mn4-z4l:-MZOTdFZ8B7WvFBc9neq:b-lvq2hj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8abc2ebb0734126824e8d1c7b279c979b5a11f40/pkg/server/api_v2_admin_test.go#L63)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This test suite feels like it should be converted to a `datadriven` test. The payloads would look good in expected test outputs.

I don't think you absolutely need to do this now (in this PR) but if you choose not to, please file an issue and outline for the benefit of the next person how this could look like.
</blockquote></details>

I'll file an issue about this, great point.

**Review by knz — COMMENTED** [review]
I'm wondering why the `cli/testdata` zip test results needed to change in this PR?

> Reviewed 9 of 9 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-:-MZSXkg0D3ZPYa-akmsE:b-rxqzto)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @itsbilal)

___
*[docs/generated/swagger/spec.json, line 53 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSXHlz70Yhw5NIojdP:-MZSXHm--ELRu1cyihci:bmllf3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L53)):*
> ```JSON
>       }
>     },
>     "/databases/{database}/": {
> ```

Maybe `{database}/id` then
___
*[docs/generated/swagger/spec.json, line 55 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSXJEN4gaEiu474R4A:-MZSXJEN4gaEiu474R4B:b-2jv2km) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L55)):*
> ```JSON
>     "/databases/{database}/": {
>       "get": {
>         "description": "Returns details about a database's descriptor ID.",
> ```

"Returns the descriptor ID of a database."

**Comment by jordanlewis** [comment]
@Azhng @maryliag can you please review this PR? We will be the eventual consumers of this API, so it's important that we take a look and understand the contracts :)

**Review by maryliag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @itsbilal)

___
*[docs/generated/swagger/spec.json, line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSmaXH8UCwOEj1KGZo:-MZSmaXH8UCwOEj1KGZp:bng5t2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L38)):*
> ```JSON
>           {
>             "type": "integer",
>             "description": "Continuation token for results after a past limited run.",
> ```

Can we not call "token" here (and all the other offset descriptions)? It makes it sounds that it's more than an integer offset (like it would need some extra info, as authentication for example)
___
*[docs/generated/swagger/spec.json, line 85 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSldtAA-QLVCAXopIC:-MZSldtAA-QLVCAXopID:bbmfl2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L85)):*
> ```JSON
>     "/databases/{database}/grants/": {
>       "get": {
>         "description": "Grants are the privileges granted to users on this database.",
> ```

nit: Returns grants on the database. Grants are...
___
*[docs/generated/swagger/spec.json, line 167 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSoi9H1OOa82YuKMvc:-MZSoi9H1OOa82YuKMvd:bd29vp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L167)):*
> ```JSON
>       }
>     },
>     "/databases/{database}/tables/{table}/": {
> ```

How does it handle having same table name on different schemas on the same database? Or do you have to pass the schema info on the table name?
___
*[docs/generated/swagger/spec.json, line 215 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSpQLR0YrdlPpXIVYs:-MZSpQLR0YrdlPpXIVYt:b-4qdiig) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L215)):*
> ```JSON
>           {
>             "type": "string",
>             "description": "Type of events to filter for (e.g. \"create_table\")",
> ```

If you want to filter by more than one type, it would be `type1,type2` or only one value per request is allowed?
___
*[docs/generated/swagger/spec.json, line 221 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSpf6C-Uq2YnTCi0s4:-MZSpf6C-Uq2YnTCi0s5:b-yzcgp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L221)):*
> ```JSON
>           {
>             "type": "integer",
>             "description": "Filter for events with this targetID.",
> ```

same as type. Can you filter by multiple targets?

**Review by dhartunian — COMMENTED** [review]
:lgtm: nice! these endpoints will greatly ease future work on the databases page

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-:-MZSqpTBCIDpiA7S1amZ:b-3wrugj)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal)

___
*[pkg/server/api_v2_sql_schema.go, line 482 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSm7WG2qyS7-OotrGa:-MZSm7WG2qyS7-OotrGb:bdjk1mw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/pkg/server/api_v2_sql_schema.go#L482)):*
> ```Go
> 	ctx := r.Context()
> 	username := security.MakeSQLUsernameFromPreNormalizedString(
> 		ctx.Value(webSessionUserKey{}).(string))
> ```

nit: I've seen this pattern hidden behind a getter function so the key type and the cast are consistent. nbd, just curious if there's a reason to avoid it.

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal)

___
*[docs/generated/swagger/spec.json, line 60 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZTOzhQChFH9x4BV64D:-MZTOzhR4iO9ZG2u88Vq:b-fdn1uw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L60)):*
> ```JSON
>         ],
>         "summary": "Get database descriptor ID",
>         "operationId": "databaseDetails",
> ```

This endpoint returns only the descriptor ID of the database. So instead of calling it `databaseDetails`, perhaps call it `databaseDescriptorID` ? 
___
*[docs/generated/swagger/spec.json, line 206 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZTQ9Fv0rEX45KQOWnL:-MZTQ9Fv0rEX45KQOWnM:b-r2jazf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L206)):*
> ```JSON
>     "/events/": {
>       "get": {
>         "description": "Lists the latest event log entries, in descending order.",
> ```

Do we want to give users the ability the specify the order ?

**Review by itsbilal — COMMENTED** [review]
Thanks for all the reviews! Really appreciate all the attention this PR has gotten lately. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-:-MZUV0W-9pJpczrqBzUM:b-2g00ok)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @dhartunian, @itsbilal, @knz, and @maryliag)

___
*[docs/generated/swagger/spec.json, line 38 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSmaXH8UCwOEj1KGZo:-MZUIpwrBjFzTUx6w8SV:b-l2m8yo) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L38)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

Can we not call "token" here (and all the other offset descriptions)? It makes it sounds that it's more than an integer offset (like it would need some extra info, as authentication for example)
</blockquote></details>

Done. Just switched to calling it an offset for the cases where it's not a base64 encoded token or something.
___
*[docs/generated/swagger/spec.json, line 53 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSXHlz70Yhw5NIojdP:-MZULsxR1bywTMQNnj4H:bbtuaq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L53)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Maybe `{database}/id` then
</blockquote></details>

Maybe we could expand the scope of this endpoint in the future when we add some field for zone config? I imagine adding new fields won't be an API-breaking change, but sunsetting paths would be, so I just kept it vague. 
___
*[docs/generated/swagger/spec.json, line 55 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSXJEN4gaEiu474R4A:-MZUMTYtAin3K-R4jKHx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L55)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"Returns the descriptor ID of a database."
</blockquote></details>

Done.
___
*[docs/generated/swagger/spec.json, line 60 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZTOzhQChFH9x4BV64D:-MZUMImn0Po-tfa-odmT:bmp8q15) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L60)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

This endpoint returns only the descriptor ID of the database. So instead of calling it `databaseDetails`, perhaps call it `databaseDescriptorID` ? 
</blockquote></details>

I replied to knz about this above, but basically, this endpoint could be expanded to include some info about the table's zone config later. 
___
*[docs/generated/swagger/spec.json, line 85 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSldtAA-QLVCAXopIC:-MZUOlwY37InIoLjDCf-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L85)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

nit: Returns grants on the database. Grants are...
</blockquote></details>

Done.
___
*[docs/generated/swagger/spec.json, line 167 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSoi9H1OOa82YuKMvc:-MZUQnkwDTOefIyPyPCG:b8ss334) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L167)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

How does it handle having same table name on different schemas on the same database? Or do you have to pass the schema info on the table name?
</blockquote></details>

Clarified. It'll try to resolve `database.tablename` if a schema wasn't specified, but you could specify a schema name to disambiguate.
___
*[docs/generated/swagger/spec.json, line 206 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZTQ9Fv0rEX45KQOWnL:-MZUSnx060JU-PgLtQtO:bmfpqnj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L206)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Do we want to give users the ability the specify the order ? 
</blockquote></details>

Do you think it'd be useful? Should be fairly straightforward to plug through.
___
*[docs/generated/swagger/spec.json, line 215 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSpQLR0YrdlPpXIVYs:-MZUU98C-IUKb21VKqrR:b-ha81be) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L215)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

If you want to filter by more than one type, it would be `type1,type2` or only one value per request is allowed?
</blockquote></details>

Clarified. Only one can be specified at once, matching apiv1 behaviour, though we can probably enhance this if it's deemed valuable.
___
*[docs/generated/swagger/spec.json, line 221 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSpf6C-Uq2YnTCi0s4:-MZUUvKb5NPCZbmcMOgR:b-m4rczg) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L221)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

same as type. Can you filter by multiple targets?
</blockquote></details>

Same as type, only one at a time. 
___
*[pkg/server/api_v2_sql_schema.go, line 482 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZSm7WG2qyS7-OotrGa:-MZUSxV44vzQb0Tvyawx:b23tj77) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/pkg/server/api_v2_sql_schema.go#L482)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: I've seen this pattern hidden behind a getter function so the key type and the cast are consistent. nbd, just curious if there's a reason to avoid it.
</blockquote></details>

Good point! Added a getter.

**Review by maryliag — COMMENTED** [review]
Great work :) :lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-:-MZUcrEuB4wY8GBVAnsC:bb8yhh6)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @dhartunian, @itsbilal, and @knz)

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian, @itsbilal, and @knz)

___
*[docs/generated/swagger/spec.json, line 60 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZTOzhQChFH9x4BV64D:-MZUdZaR7duSKqGHR_m5:bdnes1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L60)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

I replied to knz about this above, but basically, this endpoint could be expanded to include some info about the table's zone config later. 
</blockquote></details>

Sounds good. 
___
*[docs/generated/swagger/spec.json, line 206 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZTQ9Fv0rEX45KQOWnL:-MZUdafdCHssS-8Lb94y:b-t5spoa) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L206)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Do you think it'd be useful? Should be fairly straightforward to plug through.
</blockquote></details>

Would be nice to have it. I'll leave it up to you to decide :P

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian, @itsbilal, and @knz)

___
*[docs/generated/swagger/spec.json, line 39 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZXfB3v0npUJYSYobRU:-MZXfB3v0npUJYSYobRV:b-3tlvmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L39)):*
> ```JSON
>             "type": "integer",
>             "description": "Continuation token for results after a past limited run.",
>             "name": "offset",
> ```

Not this PR, but: I just realized that the "offset" abstraction is flawed.

I do believe that "token" would be better than "offset", with type string and not integer
*and the token should include a txn timestamp to use to look up the data*

The problem with the current design is that the output is not stable, it reads at the current timestamp so if the data changes between calls we get bogus results.

The proper way to do this is to have the first call compute a txn timestamp, then include that timestamp in the low-level queries (with AOST), then also embed the txn timestamp in the continuation token.

What do you want to do about this? At least filing a follow-up issue would be nice. Also I would keep the name "token" and perhaps announce the type "string" instead of "integer", so that client applications do not get the wrong impression.

**Review by itsbilal — COMMENTED** [review]
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-:-MZZ24Pn7_oYou2qG71P:bodi8iv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @dhartunian, @itsbilal, and @knz)

___
*[docs/generated/swagger/spec.json, line 39 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63000#-MZXfB3v0npUJYSYobRU:-MZZ23d67hhXt-iE_HOu:b61as6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/2736b7482cf8aeaec28afe95bb4edd7b0451441e/docs/generated/swagger/spec.json#L39)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Not this PR, but: I just realized that the "offset" abstraction is flawed.

I do believe that "token" would be better than "offset", with type string and not integer
*and the token should include a txn timestamp to use to look up the data*

The problem with the current design is that the output is not stable, it reads at the current timestamp so if the data changes between calls we get bogus results.

The proper way to do this is to have the first call compute a txn timestamp, then include that timestamp in the low-level queries (with AOST), then also embed the txn timestamp in the continuation token.

What do you want to do about this? At least filing a follow-up issue would be nice. Also I would keep the name "token" and perhaps announce the type "string" instead of "integer", so that client applications do not get the wrong impression.
</blockquote></details>

Great points! I've filed a follow-up issue about this: #64487. Also reverted back to calling it a token.

**Comment by itsbilal** [comment]
bors r=knz

**Comment by itsbilal** [comment]
Retrying due to version-upgrade failure

bors r=knz

**Comment by itsbilal** [comment]
bors retry
